proposals/0035-linalg-matrix.md

```diff
@@ -47,70 +47,816 @@ concept ArithmeticScalar = std::is_arithmetic<T>::value;
+  void
+  Store(RWByteAddressBuffer Res, uint StartOffset, uint Stride, bool ColMajor,
+        uint Align = sizeof(ElementType));
+
+  template <typename T>
+    requires ArithmeticScalar<T>
+  void Store(groupshared T Arr[], uint StartIdx, uint Stride,
+             bool ColMajor);
+
+  // Row accesses
+  vector<ElementType, M> GetRow(uint Index);
+  void SetRow(vector<ElementType, M> V, uint Index);
+
+  // Element access
+  std::enable_if_t<__detail::ComponentTypeTraits<ComponentTy>::IsNativeScalar,
+                   ElementType>
+  Get(uint2 Index);
+  std::enable_if_t<__detail::ComponentTypeTraits<ComponentTy>::IsNativeScalar,
+                   void>
+  Set(ElementType V, uint2 Index);
 
-  void Store(groupshared ElTy Arr[], uint StartIdx, uint Stride, bool ColMajor);
+  template <MatrixComponentType LHSTy, MatrixComponentType RHSTy, unit K>
+  std::enable_if_t<Use == MatrixUse::Accumulator && Scope == MatrixScope::Wave,
+                   void>
+  MultiplyAccumulate(const Matrix<LHSTy, M, K, MatrixUse::A, Scope>,
+                     const Matrix<RHSTy, K, N, MatrixUse::B, Scope>);
 
-  void MultiplyAccumulate(const ref Matrix<T, N, M>);
-  void SumAccumulate(const ref Matrix<T, N, M>);
+  template <MatrixComponentType LHSTy, MatrixComponentType RHSTy, unit K>
+  std::enable_if_t<Use == MatrixUse::Accumulator && Scope == MatrixScope::Wave,
+                   void>
+  SumAccumulate(const Matrix<LHSTy, M, K, MatrixUse::A, Scope>,
+                const Matrix<RHSTy, K, N, MatrixUse::B, Scope>);
+
+  // Cooperative Vector outer product accumulate.
+  template <typename T>
+  std::enable_if_t<Use == MatrixUse::Accumulator, void>
+  OuterProductAccumulate(const vector<T, M> &, const vector<T, N> &);
 };
 
 template <typename T, uint M, uint N, uint K>
-Matrix<T, M, N> Multiply(const ref Matrix<T, M, K>, const ref Matrix<T, K, N>);
+Matrix<T, M, N, MatrixUse::Accumulator, MatrixScope::Wave>
+
+Multiply(const Matrix<T, M, K, MatrixUse::A, MatrixScope::Wave>,
+         const Matrix<T, K, N, MatrixUse::B, MatrixScope::Wave>);
+
+// Cooperative Vector Replacement API
+// Cooperative Vector operates on per-thread vectors multiplying against B
+// matrices.
+
+template <typename OutputElTy, typename InputElTy, uint M, uint K,
+          MatrixComponentType MatrixDT, MatrixScope Scope, bool MatrixTranspose>
+vector<OutputElTy, K>
+Multiply(vector<InputElTy, M> InputVector,
+         Matrix<MatrixDT, M, K, MatrixUse::B, Scope> Matrix);
+
+template <typename OutputElTy, typename InputElTy, typename BiasElTy, uint M,
+          uint K, MatrixComponentType MatrixDT, MatrixScope Scope,
+          bool MatrixTranspose>
+vector<OutputElTy, K>
+MultiplyAdd(vector<InputElTy, M> InputVector,
+            Matrix<MatrixDT, M, K, MatrixUse::B, Scope> Matrix,
+            vector<BiasElTy, K> BiasVector);
 
 } // namespace linalg
 } // namespace hlsl
 ```
 
 ## Detailed design
 
+### HLSL API Concepts

```

Comment:
Something should mention that the API will need a way to enumerate the valid combinations of the template parameters.  AFAIK there is no single set of parameters that works everywhere, and programmers should be cautioned about that.